web/tests/ui_util.test.js

```diff
@@ -0,0 +1,80 @@
+"use strict";
+
+const {strict: assert} = require("assert");
+
+const {zrequire} = require("./lib/namespace");
+const {run_test} = require("./lib/test");
+const $ = require("./lib/zjquery");
+
+const ui_util = zrequire("ui_util");
+
+run_test("potentially_collapse_quotes", ({override_rewire}) => {
+    const $element = $.create("message-content");
+    let children = [];
+    $element.children = () => children;
+
+    children = [
+        $.create("normal paragraph 1"),
+        $.create("blockquote"),
+        $.create("normal paragraph 2"),
+        $.create("user said paragraph"),
+        $.create("user quote"),
+        $.create("normal paragraph 3"),
+    ];
+    override_rewire(ui_util, "get_collapsible_status_array", () => [
+        false,
+        true,
+        false,
+        true,
+        true,
+        false,
+    ]);
+    // When there are both collapsible and non-collapsible elements, for
+    // multiple collapsible elements in a row, only the first element
+    // should be collapsed, and the rest's text should be removed. Non-
+    // collapsible elements should not be touched.
+    let collapsed = ui_util.potentially_collapse_quotes($element);
+    assert.equal(collapsed, true);
+    let expected_texts = [
+        "never-been-set",
+        "[...]",
+        "never-been-set",
+        "[...]",
+        "",
+        "never-been-set",
+    ];
+    assert.deepEqual(
+        $element.children().map(($el) => $el.text()),
+        expected_texts,
+    );
+
+    children = [
+        $.create("normal paragraph 4"),
+        $.create("normal paragraph 5"),
+        $.create("normal paragraph 6"),
+    ];
+    override_rewire(ui_util, "get_collapsible_status_array", () => [false, false, false]);
+    // For all non-collapsible elements, none should be collapsed.
+    collapsed = ui_util.potentially_collapse_quotes($element);
+    assert.equal(collapsed, false);
+    expected_texts = ["never-been-set", "never-been-set", "never-been-set"];
+    assert.deepEqual(
+        $element.children().map(($el) => $el.text()),
+        expected_texts,
+    );
+
+    children = [$.create("blockquote 1"), $.create("blockquote 2"), $.create("blockquote 3")];
+    override_rewire(ui_util, "get_collapsible_status_array", () => [true, true, true]);
+    // For all collapsible elements, none should be collapsed.
+    collapsed = ui_util.potentially_collapse_quotes($element);
+    assert.equal(collapsed, false);
+    expected_texts = ["never-been-set", "never-been-set", "never-been-set"];
+    assert.deepEqual(
+        $element.children().map(($el) => $el.text()),
+        expected_texts,
+    );
+});
+
+run_test("is_user_said_paragraph", () => {
+    // yet to figure this out
```

Comment:
@timabbott The `is_user_said_paragraph` function involves a lot of complex DOM parsing using a variety of jQuery functions, which aren't available in zjquery, and are also not easy to mock without essentially rewriting the logic which renders the test useless... Given that we do have tests for how collapsing works once we know what's to be collapsed, should we skip the test suite for this function?

Reply:
Yeah, it's not worth writing tests that are hard to write and have limited value.

Reply:
Removed the stub, this is ready for another review now!